### ISSUE TITLE
Add a rule to flag non-symbolic function-invoked operators

### ISSUE THREAD

**Issue by veyndan** [detekt/detekt#3437]
Add a rule to flag non-symbolic function-invoked operators (e.g., `plus`, `minus`) to promote the usage of it's symbolic counterparts (e.g., `+`, `-`). It'd also be useful if we could ignore this rule for certain operators (e.g., configuring `invoke` to be ignored would mean calling `foo.invoke()` would pass this rule).

## Expected Behavior of the rule

The below can also be applied to the other non-symbolic function-invoked operators.

### Noncompliant Code

```kotlin
val result = 2.plus(2)
assert(result.equals(4))
```

### Compliant Code

```kotlin
val result = 2 + 2
assert(result == 4)
```

**Comment by cortinico** [comment]
I think that's a valid addition but there are couple of points that could make this rule controversial:

* There are valid use cases for function-invoke operators (e.g. if you're chaining with a nullable type: `a?.plus`)
* The example that you mentioned with `.equals` is misleading as the `a == b` operator calls translates to `a?.equals(b) ?: (b === null)` ([source](https://kotlinlang.org/docs/operator-overloading.html#equality-and-inequality-operators))

**Comment by schalkms** [comment]
I also think this rule would be very controversial. I don't see why the usage of `plus` instead of `+` would be a bad thing or to put it in other words why the operator style should be promoted. I don't think this is a style related rule.

**Comment by BraisGabin** [comment]
I prefer to use `list.plus(element)` than `list + element`. I agree that the usage of + for integers is better but I don't know if we need a rule for that.

**Comment by veyndan** [comment]
> I think that's a valid addition but there are couple of points that could make this rule controversial:
> 
> * There are valid use cases for function-invoke operators (e.g. if you're chaining with a nullable type: `a?.plus`)
> * The example that you mentioned with `.equals` is misleading as the `a == b` operator calls translates to `a?.equals(b) ?: (b === null)` ([source](https://kotlinlang.org/docs/operator-overloading.html#equality-and-inequality-operators))

Yeah that makes sense. Maybe we could make it such that method references and safe calls are the exception?

> I also think this rule would be very controversial. I don't see why the usage of `plus` instead of `+` would be a bad thing or to put it in other words why the operator style should be promoted. I don't think this is a style related rule.

@schalkms From @cortinico (and correct me if I'm wrong), they said that it would be controversial based on how something like `?.plus` would be handled, not on the rule itself.

> I prefer to use `list.plus(element)` than `list + element`. I agree that the usage of + for integers is better but I don't know if we need a rule for that.

@BraisGabin I think that this is a special case based on your personal style guidelines. You could easily ignore `plus` in the proposed rule, and then you can use `plus` or `+` interchangeably (or enforce it by creating your own custom rule).

> > I also think this rule would be very controversial.
> 
> But isn't exactly the point that you could simply disable then rule then? Or _detekt_ could even disable it by default?
> 
> But to enforce project-specific coding conventions such a rule would be very handy. Personally, I strongly prefer e.g. the use of `+=` over `.add()`, and `in` over `.contains()` etc.

I agree with @sschuberth, there are several rules that could be considered controversial, but the fact that they are _opt-in_ makes it fine. If we were requiring this rule to be enabled for everyone, I could see the concern, but this clearly isn't the case in detekt.

Our team promotes (and we'd like to enforce) that all operators should use the symbolic notation and not the function name (apart from a few functions, e.g., `invoke`) and allow it in method references and safe calls. I believe that this is a common enough code style to warrant being added to the detekt codebase.

**Comment by cortinico** [comment]
> I agree with @sschuberth, there are several rules that could be considered controversial, but the fact that they are _opt-in_ makes it fine. If we were requiring this rule to be enabled for everyone, I could see the concern, but this clearly isn't the case in detekt.

Agree that this can be added as disabled by default, so users can opt-in and enable it if they wish. Before we get into implementation mode, can I ask you to clarify which exact scenarios should the rule exclude/include, or which config flags are you suggesting?

**Comment by cortinico** [comment]
Do we want this rule in Detekt? If yes, I think this is up for grab (will label it as "help wanted") otherwise I'm fine closing it.

**Comment by sschuberth** [comment]
I *do* like to see this rule in *detekt*, but I currently have no time to work on it.

**Comment by BraisGabin** [comment]
> Before we get into implementation mode, can I ask you to clarify which exact scenarios should the rule exclude/include, or which config flags are you suggesting?

I agree with @cortinico on this. I would like to know the scope and how configurable this rule will be. I can imagine some configurations:

- List of operators that should be used in the symbolic way. (Or list of operators that should be excluded?)
- List of Types where this rule applies. (Or list of types to ignore)
- A matrix for those two. (Probably that would be far too much)

**Comment by sschuberth** [comment]
IMO, we should keep it simple and, if the rule is enabled, point out *any* function call where an overloaded operator / symbolic function invocation could be used. To me, this rule is mainly a reminder for people coming from Java, where operator overloading is not possible.
